### PR TITLE
Kubernetes Data Protection Extension CLI

### PR DESCRIPTION
**Pull request by Miraj50**
**Data Protection Kubernetes Extension**
We are authoring a new k8s partner extension for our BCDR solution of AKS clusters.
This PR consists of code that 
- Takes in input parameters from the user and converts it to the format required for our helm chart
- Do the required checks before the create/update of the extension

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repository and upgrade the version in the pull request but do not modify `src/index.json`.

### CODE REVIEW

File: src/k8s-extension/azext_k8s_extension/partner_extensions/DataProtectionKubernetes.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        tenant_id = self.__get_tenant_id(cmd.cli_ctx)
+        if not tenant_id:
+            raise SystemExit(logger.error("Unable to fetch TenantId. Please check your subscription or run 'az login' to login to Azure."))
+
+        self.__validate_and_map_config(configuration_settings)
+        self.__validate_backup_storage_account(cmd.cli_ctx, resource_group_name, cluster_name, configuration_settings)
+
+        configuration_settings[self.TENANT_ID] = tenant_id
+
+        if release_train is None:
```
Comment (anshulahuja98) [inline src/k8s-extension/azext_k8s_extension/partner_extensions/DataProtectionKubernetes.py:99]: Default should be prod right?

File: src/k8s-extension/azext_k8s_extension/partner_extensions/DataProtectionKubernetes.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        self.BACKUP_STORAGE_ACCOUNT_RESOURCE_GROUP = "configuration.backupStorageLocation.config.resourceGroup"
+        self.BACKUP_STORAGE_ACCOUNT_SUBSCRIPTION = "configuration.backupStorageLocation.config.subscriptionId"
+        self.RESOURCE_LIMIT_CPU = "resources.limits.cpu"
+        self.RESOURCE_LIMIT_MEMORY = "resources.limits.memory"
+
+        self.blob_container = "blobContainer"
+        self.storage_account = "storageAccount"
+        self.storage_account_resource_group = "storageAccountResourceGroup"
+        self.storage_account_subsciption = "storageAccountSubscriptionId"
+        self.cpu_limit = "cpu"
```
Comment (anshulahuja98) [inline src/k8s-extension/azext_k8s_extension/partner_extensions/DataProtectionKubernetes.py:37]: rename to cpuLimit
and memoryLimit resp?

**Review by anshulahuja98 — CHANGES_REQUESTED** [review]

File: src/k8s-extension/azext_k8s_extension/_client_factory.py [hunk 3]
```diff
@@ -51,3 +51,13 @@ def cf_log_analytics(cli_ctx, subscription_id=None):
...
     return get_mgmt_service_client(cli_ctx, ResourceManagementClient).providers
+
+
+def cf_storage(cli_ctx, subscription_id=None):
+    from azure.mgmt.storage import StorageManagementClient
+    return get_mgmt_service_client(cli_ctx, StorageManagementClient, subscription_id=subscription_id)
+
+
+def cf_managed_clusters(cli_ctx, subscription_id=None):
+    from azure.mgmt.containerservice import ContainerServiceClient
```
[the remark this answers is not in the document]
Comment (Miraj50) [reply src/k8s-extension/azext_k8s_extension/_client_factory.py:62]: We need the client to do a GET on the cluster to get its location so we can verify if the cluster and SA are in the same location. Other extensions don't do a GET.

File: src/k8s-extension/azext_k8s_extension/partner_extensions/DataProtectionKubernetes.py (hunk 1 above)
Comment (Miraj50) [reply src/k8s-extension/azext_k8s_extension/partner_extensions/DataProtectionKubernetes.py:99]: Or 'stable'?

File: src/k8s-extension/azext_k8s_extension/partner_extensions/DataProtectionKubernetes.py (hunk 2 above)
Comment (Miraj50) [reply src/k8s-extension/azext_k8s_extension/partner_extensions/DataProtectionKubernetes.py]: Done

File: src/k8s-extension/azext_k8s_extension/partner_extensions/DataProtectionKubernetes.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+        self.__validate_and_map_config(configuration_settings)
+        self.__validate_backup_storage_account(cmd.cli_ctx, resource_group_name, cluster_name, configuration_settings)
+
+        configuration_settings[self.TENANT_ID] = tenant_id
+
+        if release_train is None:
+            release_train = 'stable'
+
+        create_identity = True
```
Comment (bavneetsingh16) [inline src/k8s-extension/azext_k8s_extension/partner_extensions/DataProtectionKubernetes.py:102]: I believe we can set this flag to false as identities are not created for managedClusters and if that is the only clusterType that will be supported and if it is not about to change.

[https://github.com/AzureArcForKubernetes/azure-cli-extensions/blob/k8s-extension/public/src/k8s-extension/azext_k8s_extension/custom.py#L416](https://github.com/AzureArcForKubernetes/azure-cli-extensions/blob/k8s-extension/public/src/k8s-extension/azext_k8s_extension/custom.py#L416)

Comment (bavneetsingh16) [reply src/k8s-extension/azext_k8s_extension/partner_extensions/DataProtectionKubernetes.py:102]: nvm, since we already have the logic present for AKS

**Review by bavneetsingh16 — APPROVED** [review]
LGTM

File: src/k8s-extension/azext_k8s_extension/partner_extensions/DataProtectionKubernetes.py (hunk 4 above)
Comment (Miraj50) [reply src/k8s-extension/azext_k8s_extension/partner_extensions/DataProtectionKubernetes.py:102]: Got it Bavneet!

**Pull request merged**